File: BackEnd/PHP/laravelshop/app/Modules/Community/Services/CommunityService.php
<?php

namespace App\Modules\Community\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\Board;
use App\Models\Post;
use App\Models\PostComment;
use App\Models\PostLike;
use App\Models\User;
use Illuminate\Support\Facades\DB;
use Symfony\Component\HttpFoundation\Response;

class CommunityService
{
    public function listBoards(): array
    {
        return Board::query()
            ->withCount('posts')
            ->orderBy('id')
            ->get()
            ->map(fn (Board $board): array => [
                'id' => $board->id,
                'name' => $board->name,
                'slug' => $board->slug,
                'description' => $board->description,
                'postCount' => $board->posts_count,
                'createdAt' => optional($board->created_at)?->toISOString(),
                'updatedAt' => optional($board->updated_at)?->toISOString(),
            ])
            ->values()
            ->all();
    }

    public function listPosts(int $boardId): array
    {
        $board = Board::query()->find($boardId);
        if (! $board) {
            throw new BusinessException('게시판을 찾을 수 없습니다.', 'BOARD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $items = Post::query()
            ->with(['board', 'user'])
            ->where('board_id', $boardId)
            ->orderByDesc('id')
            ->get()
            ->map(fn (Post $post): array => $this->serializePost($post))
            ->values()
            ->all();

        return [
            'board' => [
                'id' => $board->id,
                'name' => $board->name,
                'slug' => $board->slug,
            ],
            'items' => $items,
        ];
    }

    public function showPost(int $postId): array
    {
        $post = Post::query()->with(['board', 'user'])->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $post->increment('view_count');
        $post->refresh()->load(['board', 'user']);

        return $this->serializePost($post);
    }

    public function createPost(User $user, int $boardId, array $payload): array
    {
        $board = Board::query()->find($boardId);
        if (! $board) {
            throw new BusinessException('게시판을 찾을 수 없습니다.', 'BOARD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $post = Post::query()->create([
            'board_id' => $boardId,
            'user_id' => $user->id,
            'title' => $payload['title'],
            'content' => $payload['content'],
        ]);

        return $this->serializePost($post->fresh(['board', 'user']));
    }

    public function updatePost(User $actor, int $postId, array $payload): array
    {
        $post = Post::query()->with(['board', 'user'])->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }
        if ($actor->role !== 'ADMIN' && $post->user_id !== $actor->id) {
            throw new BusinessException('게시글 수정 권한이 없습니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }

        $updates = [];
        if (array_key_exists('title', $payload)) {
            $updates['title'] = $payload['title'];
        }
        if (array_key_exists('content', $payload)) {
            $updates['content'] = $payload['content'];
        }

        if ($updates !== []) {
            $post->forceFill($updates)->save();
        }

        return $this->serializePost($post->fresh(['board', 'user']));
    }

    public function deletePost(User $actor, int $postId): array
    {
        $post = Post::query()->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }
        if ($actor->role !== 'ADMIN' && $post->user_id !== $actor->id) {
            throw new BusinessException('게시글 삭제 권한이 없습니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }

        $post->delete();

        return ['message' => '게시글이 삭제되었습니다.'];
    }

    public function toggleLike(User $user, int $postId): array
    {
        $post = Post::query()->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $like = PostLike::query()->where('post_id', $postId)->where('user_id', $user->id)->first();
        $liked = false;

        if ($like) {
            $like->delete();
        } else {
            PostLike::query()->create([
                'post_id' => $postId,
                'user_id' => $user->id,
            ]);
            $liked = true;
        }

        $likeCount = PostLike::query()->where('post_id', $postId)->count();
        $post->forceFill(['like_count' => $likeCount])->save();

        return [
            'postId' => $postId,
            'liked' => $liked,
            'likeCount' => $likeCount,
        ];
    }

    public function listComments(int $postId): array
    {
        $post = Post::query()->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return PostComment::query()
            ->with('user')
            ->where('post_id', $postId)
            ->orderBy('id')
            ->get()
            ->map(fn (PostComment $comment): array => $this->serializeComment($comment))
            ->values()
            ->all();
    }

    public function createComment(User $user, int $postId, array $payload): array
    {
        $post = Post::query()->find($postId);
        if (! $post) {
            throw new BusinessException('게시글을 찾을 수 없습니다.', 'POST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $comment = DB::transaction(function () use ($user, $post, $payload): PostComment {
            $comment = PostComment::query()->create([
                'post_id' => $post->id,
                'user_id' => $user->id,
                'content' => $payload['content'],
            ]);

            $post->increment('comment_count');

            return $comment;
        });

        return $this->serializeComment($comment->fresh('user'));
    }

    public function deleteComment(User $actor, int $commentId): array
    {
        $comment = PostComment::query()->find($commentId);
        if (! $comment) {
            throw new BusinessException('댓글을 찾을 수 없습니다.', 'COMMENT_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }
        if ($actor->role !== 'ADMIN' && $comment->user_id !== $actor->id) {
            throw new BusinessException('댓글 삭제 권한이 없습니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }

        DB::transaction(function () use ($comment): void {
            Post::query()->whereKey($comment->post_id)->where('comment_count', '>', 0)->decrement('comment_count');
            $comment->delete();
        });

        return ['message' => '댓글이 삭제되었습니다.'];
    }

    private function serializePost(Post $post): array
    {
        return [
            'id' => $post->id,
            'board' => $post->board ? [
                'id' => $post->board->id,
                'name' => $post->board->name,
                'slug' => $post->board->slug,
            ] : null,
            'title' => $post->title,
            'content' => $post->content,
            'viewCount' => $post->view_count,
            'likeCount' => $post->like_count,
            'commentCount' => $post->comment_count,
            'author' => $post->user ? [
                'id' => $post->user->id,
                'name' => $post->user->name,
                'nickname' => $post->user->nickname,
            ] : null,
            'createdAt' => optional($post->created_at)?->toISOString(),
            'updatedAt' => optional($post->updated_at)?->toISOString(),
        ];
    }

    private function serializeComment(PostComment $comment): array
    {
        return [
            'id' => $comment->id,
            'postId' => $comment->post_id,
            'content' => $comment->content,
            'author' => $comment->user ? [
                'id' => $comment->user->id,
                'name' => $comment->user->name,
                'nickname' => $comment->user->nickname,
            ] : null,
            'createdAt' => optional($comment->created_at)?->toISOString(),
            'updatedAt' => optional($comment->updated_at)?->toISOString(),
        ];
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/community/CommunityService.java
package com.pbshop.java.spring.maven.jpa.postgresql.community;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.community.dto.CommunityDtos.SaveCommentRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.community.dto.CommunityDtos.SavePostRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.user.User;
import com.pbshop.java.spring.maven.jpa.postgresql.user.UserRepository;

@Service
@Transactional(readOnly = true)
public class CommunityService {

    private final BoardRepository boardRepository;
    private final PostRepository postRepository;
    private final PostCommentRepository postCommentRepository;
    private final PostLikeRepository postLikeRepository;
    private final UserRepository userRepository;

    public CommunityService(
            BoardRepository boardRepository,
            PostRepository postRepository,
            PostCommentRepository postCommentRepository,
            PostLikeRepository postLikeRepository,
            UserRepository userRepository
    ) {
        this.boardRepository = boardRepository;
        this.postRepository = postRepository;
        this.postCommentRepository = postCommentRepository;
        this.postLikeRepository = postLikeRepository;
        this.userRepository = userRepository;
    }

    public List<Map<String, Object>> getBoards() {
        return boardRepository.findByActiveTrueOrderByIdAsc().stream()
                .map(this::toBoardResponse)
                .toList();
    }

    public Map<String, Object> getBoardPosts(Long boardId, String search, String sort, int page, int limit) {
        Board board = getBoard(boardId);
        Pageable pageable = PageRequest.of(Math.max(page - 1, 0), limit, resolveSort(sort));
        Page<Post> postPage = postRepository.findBoardPosts(boardId, search, pageable);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("board", toBoardResponse(board));
        response.put("items", postPage.getContent().stream().map(this::toPostSummary).toList());
        response.put("page", postPage.getNumber() + 1);
        response.put("limit", postPage.getSize());
        response.put("total", postPage.getTotalElements());
        response.put("totalPages", postPage.getTotalPages());
        return response;
    }

    @Transactional
    public Map<String, Object> getPost(Long postId) {
        Post post = getPostEntity(postId);
        post.setViewCount(post.getViewCount() + 1);
        postRepository.save(post);

        Map<String, Object> response = new LinkedHashMap<>(toPostSummary(post));
        response.put("content", post.getContent());
        response.put("comments", getComments(postId));
        return response;
    }

    @Transactional
    public Map<String, Object> createPost(AuthenticatedUserPrincipal principal, Long boardId, SavePostRequest request) {
        Board board = getBoard(boardId);
        User user = getUser(principal.userId());

        Post post = new Post();
        post.setBoard(board);
        post.setUser(user);
        post.setTitle(request.title());
        post.setContent(request.content());
        post.setStatus("ACTIVE");
        return toPostSummary(postRepository.save(post));
    }

    @Transactional
    public Map<String, Object> updatePost(AuthenticatedUserPrincipal principal, Long postId, SavePostRequest request) {
        Post post = getPostEntity(postId);
        requireOwnerOrAdmin(principal, post.getUser().getId());
        post.setTitle(request.title());
        post.setContent(request.content());
        return toPostSummary(postRepository.save(post));
    }

    @Transactional
    public Map<String, Object> deletePost(AuthenticatedUserPrincipal principal, Long postId) {
        Post post = getPostEntity(postId);
        requireOwnerOrAdmin(principal, post.getUser().getId());
        postRepository.delete(post);
        return Map.of("message", "게시글이 삭제되었습니다.");
    }

    @Transactional
    public Map<String, Object> toggleLike(AuthenticatedUserPrincipal principal, Long postId) {
        Post post = getPostEntity(postId);
        return postLikeRepository.findByPostIdAndUserId(postId, principal.userId())
                .map(existing -> {
                    postLikeRepository.delete(existing);
                    post.setLikeCount((int) postLikeRepository.countByPostId(postId));
                    postRepository.save(post);
                    return Map.<String, Object>of("liked", false, "likeCount", post.getLikeCount());
                })
                .orElseGet(() -> {
                    PostLike like = new PostLike();
                    like.setPost(post);
                    like.setUser(getUser(principal.userId()));
                    postLikeRepository.save(like);
                    post.setLikeCount((int) postLikeRepository.countByPostId(postId));
                    postRepository.save(post);
                    return Map.<String, Object>of("liked", true, "likeCount", post.getLikeCount());
                });
    }

    public List<Map<String, Object>> getComments(Long postId) {
        ensurePostExists(postId);
        return postCommentRepository.findByPostIdOrderByIdAsc(postId).stream()
                .map(this::toCommentResponse)
                .toList();
    }

    @Transactional
    public Map<String, Object> createComment(AuthenticatedUserPrincipal principal, Long postId, SaveCommentRequest request) {
        Post post = getPostEntity(postId);
        PostComment comment = new PostComment();
        comment.setPost(post);
        comment.setUser(getUser(principal.userId()));
        comment.setContent(request.content());
        PostComment saved = postCommentRepository.save(comment);
        post.setCommentCount((int) postCommentRepository.countByPostId(postId));
        postRepository.save(post);
        return toCommentResponse(saved);
    }

    @Transactional
    public Map<String, Object> deleteComment(AuthenticatedUserPrincipal principal, Long commentId) {
        PostComment comment = postCommentRepository.findById(commentId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "댓글을 찾을 수 없습니다."));
        requireOwnerOrAdmin(principal, comment.getUser().getId());
        Long postId = comment.getPost().getId();
        postCommentRepository.delete(comment);
        Post post = getPostEntity(postId);
        post.setCommentCount((int) postCommentRepository.countByPostId(postId));
        postRepository.save(post);
        return Map.of("message", "댓글이 삭제되었습니다.");
    }

    private Board getBoard(Long boardId) {
        return boardRepository.findById(boardId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "게시판을 찾을 수 없습니다."));
    }

    private Post getPostEntity(Long postId) {
        return postRepository.findById(postId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "게시글을 찾을 수 없습니다."));
    }

    private void ensurePostExists(Long postId) {
        if (!postRepository.existsById(postId)) {
            throw new BusinessException(ErrorCode.NOT_FOUND, "게시글을 찾을 수 없습니다.");
        }
    }

    private User getUser(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }

    private void requireOwnerOrAdmin(AuthenticatedUserPrincipal principal, Long ownerId) {
        if ("ADMIN".equalsIgnoreCase(principal.role())) {
            return;
        }
        if (!principal.userId().equals(ownerId)) {
            throw new BusinessException(ErrorCode.FORBIDDEN);
        }
    }

    private Sort resolveSort(String sort) {
        if ("popular".equalsIgnoreCase(sort)) {
            return Sort.by(Sort.Direction.DESC, "likeCount").and(Sort.by(Sort.Direction.DESC, "id"));
        }
        if ("most_commented".equalsIgnoreCase(sort)) {
            return Sort.by(Sort.Direction.DESC, "commentCount").and(Sort.by(Sort.Direction.DESC, "id"));
        }
        return Sort.by(Sort.Direction.DESC, "id");
    }

    private Map<String, Object> toBoardResponse(Board board) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", board.getId());
        response.put("name", board.getName());
        response.put("slug", board.getSlug());
        response.put("description", board.getDescription());
        response.put("active", board.isActive());
        return response;
    }

    private Map<String, Object> toPostSummary(Post post) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", post.getId());
        response.put("boardId", post.getBoard().getId());
        response.put("boardName", post.getBoard().getName());
        response.put("userId", post.getUser().getId());
        response.put("userName", post.getUser().getName());
        response.put("title", post.getTitle());
        response.put("status", post.getStatus());
        response.put("viewCount", post.getViewCount());
        response.put("likeCount", post.getLikeCount());
        response.put("commentCount", post.getCommentCount());
        response.put("createdAt", post.getCreatedAt());
        return response;
    }

    private Map<String, Object> toCommentResponse(PostComment comment) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", comment.getId());
        response.put("postId", comment.getPost().getId());
        response.put("userId", comment.getUser().getId());
        response.put("userName", comment.getUser().getName());
        response.put("content", comment.getContent());
        response.put("createdAt", comment.getCreatedAt());
        return response;
    }
}
